### ISSUE TITLE
Set up a CI server to run the "short" test list

### ISSUE THREAD

**Issue by billsacks** [ESCOMP/ctsm#278]
@jhamman suggested in #275 

> Would it be possible at some point to get a subset of the CTSM tests that can be run on an arbitrary UNIX machine and to have those test be run on a Continuous Integration service like TravisCI or CircleCI? Just something to consider going forward, this tends to be standard for open-source projects these days and really helps maintain portability and usability of the model. A starting point may just be to build a standard compset with generic libraries.

**Comment by billsacks** [comment]
Looking briefly into what would be needed to get this running on Travis-CI. I initially identified two issues:

1. Fortran support: Fortran is not one of the supported languages, at least according to the official list. At first I thought we may need to add support ourselves (see https://docs.travis-ci.com/user/languages/community-supported-languages/), but I see that there are ways to get this to work, which seem to involve installing gfortran as part of your build (https://github.com/travis-ci-tester/travis-test-fortran/blob/master/.travis.yml and https://github.com/scivision/lowtran/blob/master/.travis.yml ).

2. How should we handle build dependencies like NetCDF? Do we need to completely install NetCDF (etc.) for every Travis-CI build? That could be time-consuming. Or would it make sense to create a Docker image or something similar?

**Comment by billsacks** [comment]
@martynpclark yes, very useful - thanks! And that suggests that "fortran" is an understood language even though it doesn't appear in the Travis-CI documentation.

**Comment by jhamman** [comment]
@billsacks - before you get too far down the road of TravisCI, it may be worth looking at CircleCI a bit too. It may allow you to avoid jumping through some of the hoops the TravisCI requires of Fortran/netcdf/etc. users. That said, we have gotten travis to work with SUMMA and [VIC](https://github.com/UW-Hydro/VIC/blob/master/.travis.yml), both of which have similar dependencies as CTSM.

**Comment by billsacks** [comment]
Thanks, @jhamman . I'll take a look at CircleCI when I get a chance. I'm not actively working on this yet, but was looking into TravisCI yesterday for another project (with python-only testing), so wanted to jot down some notes from that.

**Comment by billsacks** [comment]
A few notes from a discussion with @jhamman today (thanks for the discussion, Joe!)

- For VIC, they use TravisCI. They download and build the necessary dependencies (mpich, netcdf)

- CircleCI provides more flexibility in terms of docker containers if we want to go that route rather than setting things up on a fresh system

- Because of the limited time you have for running tests, we may want to have our CI tests run on a very small domain (e.g., a few grid cells) in order to reduce the time spent downloading data. The main point would be to make sure the code builds and runs.

**Comment by billsacks** [comment]
I talked with Davide Del Vento in CISL about this, and he talked with some others in CISL who are trying to support cloud computing across NCAR. Their feeling is that we might be best off doing this on one of the big cloud computing platforms like Amazon, etc. The advantage of that is that we could have some permanent disk space, so we wouldn't have to deal with rebuilding dependencies (or bringing them in via a container) for each job. (They also suggest keeping the jobs to a single node, since performance goes down substantially on these platforms when going to multiple nodes.)

It's unclear whether there's something like TravisCI for the Amazon cloud - i.e., with tie-ins to github - but Davide said he could help investigate that. I also asked: if we're going to use something more custom rather than just using something like TravisCI, would it make sense to run the jobs on our local machines rather than doing it in the cloud? His feeling was that it's likely we can find something off-the-shelf for doing this on the Amazon cloud or one of the other big cloud providers, whereas we'd have to write & maintain more ourselves if we wanted to run on our local machines. (I could also see authentication and security being an issue on our local machines.)

**Comment by ekluzek** [comment]
With CIRRUS now a thing that can support this, I think this is something that we'll be working on. And I understand that @samsrabin has funding to work on this through the CSF (Community Software Facility). So I assigned this to him, as I think that describes part of what he'll be doing there.
